components/raftstore/src/store/worker/read.rs

```diff
@@ -987,5 +1036,86 @@ mod tests {
             reader.metrics.rejected_by_term_mismatch,
             previous_term_rejection + 1,
         );
+        assert_eq!(reader.metrics.rejected_by_cache_miss, 4);
+
+        // Stale local ReadDelegate
+        cmd.mut_header().mut_region_epoch().set_version(4);
+        region1.mut_region_epoch().set_version(4);
+        let pg = Progress::region(region1);
+        {
+            let mut meta = store_meta.lock().unwrap();
+            meta.readers.get_mut(&1).unwrap().update(pg);
+        }
+        let task = RaftCommand::<KvTestSnapshot>::new(cmd, Callback::Read(Box::new(move |_| {})));
+        must_not_redirect(&mut reader, &rx, task);
+        assert_eq!(reader.metrics.rejected_by_cache_miss, 5);
+    }
+
+    #[test]
+    fn test_read_delegate_cache_update() {
+        let store_id = 2;
+        let store_meta = Arc::new(Mutex::new(StoreMeta::new(0)));
+        let (_tmp, mut reader, _) = new_reader("test-local-reader", store_id, store_meta.clone());
+        let mut region = metapb::Region::default();
+        region.set_id(1);
+        {
+            let mut meta = store_meta.lock().unwrap();
+            let read_delegate = ReadDelegate {
+                tag: String::new(),
+                region: Arc::new(region.clone()),
+                peer_id: 1,
+                term: 1,
+                applied_index_term: 1,
+                leader_lease: None,
+                last_valid_ts: Timespec::new(0, 0),
+                invalid: Arc::new(AtomicBool::new(false)),
+                txn_extra_op: Arc::new(AtomicCell::new(TxnExtraOp::default())),
+                max_ts_sync_status: Arc::new(AtomicU64::new(0)),
+                track_ver: TrackVer::new(),
+            };
+            meta.readers.insert(1, read_delegate);
+        }
+
+        let d = reader.get_delegate(1).unwrap();
+        assert_eq!(&*d.region, &region);
+        assert_eq!(d.term, 1);
+        assert_eq!(d.applied_index_term, 1);
+        assert!(d.leader_lease.is_none());
+        drop(d);
+
+        {
+            region.mut_region_epoch().set_version(10);
+            let mut meta = store_meta.lock().unwrap();
+            meta.readers
+                .get_mut(&1)
+                .unwrap()
+                .update(Progress::region(region.clone()));
+        }
+        assert_eq!(&*reader.get_delegate(1).unwrap().region, &region);
+
+        {
+            let mut meta = store_meta.lock().unwrap();
+            meta.readers.get_mut(&1).unwrap().update(Progress::term(2));
+        }
+        assert_eq!(reader.get_delegate(1).unwrap().term, 2);
+
+        {
+            let mut meta = store_meta.lock().unwrap();
+            meta.readers
+                .get_mut(&1)
+                .unwrap()
+                .update(Progress::applied_index_term(2));
+        }
+        assert_eq!(reader.get_delegate(1).unwrap().applied_index_term, 2);
+
+        {
```

Comment:
I don't think you need to put them in brackets. Now compiler is smart enough to infer the correct lifetime.

Reply:
I revert the change, remove the bracket will make the test blocking, because `reader.get_delegate` may also try to `store_meta.lock()`